libs/tdigest/src/main/java/org/elasticsearch/tdigest/AVLTreeDigest.java

```diff
@@ -0,0 +1,441 @@
+                        double partialWeight = (x - aMean) / (bMean - aMean) * (aWeight + bWeight) / 2.0;
+                        return (weightSoFar + aWeight / 2.0 + partialWeight) / size();
+                    }
+                }
+                weightSoFar += aWeight;
+
+                if (it.hasNext()) {
+                    aMean = bMean;
+                    aWeight = bWeight;
+
+                    b = it.next();
+                    bMean = b.mean();
+                    bWeight = b.count();
+
+                    assert bMean >= aMean;
+                } else {
+                    bWeight = 0;
+                }
+            }
+            // shouldn't be possible because x <= lastMean
+            throw new IllegalStateException("Ran out of centroids");
+        }
+    }
+
+    private double interpolateTail(AVLGroupTree values, double x, int node, double mean, double extremeValue) {
+        int count = values.count(node);
+        assert count > 1;
+        if (count == 2) {
+            // other sample must be on the other side of the mean
+            return 1.0 / size();
+        } else {
+            // how much weight is available for interpolation?
+            double weight = count / 2.0 - 1;
+            // how much is between min and here?
+            double partialWeight = (extremeValue - x) / (extremeValue - mean) * weight;
+            // account for sample at min along with interpolated weight
+            return (partialWeight + 1.0) / size();
+        }
+    }
+
+    /**
+     * @param q The quantile desired.  Can be in the range [0,1].
+     * @return The minimum value x such that we think that the proportion of samples is &le; x is q.
+     */
+    @Override
+    public double quantile(double q) {
+        if (q < 0 || q > 1) {
+            throw new IllegalArgumentException("q should be in [0,1], got " + q);
+        }
+
+        AVLGroupTree values = summary;
+        if (values.size() == 0) {
+            // no centroids means no data, no way to get a quantile
+            return Double.NaN;
+        } else if (values.size() == 1) {
+            // with one data point, all quantiles lead to Rome
+            return values.iterator().next().mean();
+        }
+
+        // if values were stored in a sorted array, index would be the offset we are interested in
+        final double index = q * (count - 1);
+
+        // deal with min and max as a special case singletons
+        if (index <= 0) {
+            return min;
+        }
+
+        if (index >= count - 1) {
+            return max;
+        }
+
+        int currentNode = values.first();
+        int currentWeight = values.count(currentNode);
+
+        // weightSoFar represents the total mass to the left of the center of the current node.
+        // If there's a singleton on the left boundary, there's no more mass on the left.
+        double weightSoFar = (currentWeight == 1) ? 0 : currentWeight / 2.0;
```

Comment:
I can't manage to convince myself that this is correct, it feels like it should be `currentWeight/2` even when `currentWeight` is 1, given how we then sum it up with `dw`? (I could be wrong, I haven't looked at this in a very long time)

Reply:
My understanding is that every centroid divides its weight between left and right, so about half of the samples it represents are to its left and half to its right. However, if there's a singleton on the leftmost centroid, there should be no other weight to its left. I think this is what weightSoFar represents, but I may be wrong here. `dw` extrapolates between the first and the second centroid, so the first centroid's weight (1) will still be taken into account? 

I also tried to match the results with `Dist.quantile` that uses a sorted array of all the samples. This is how I noticed this discrepancy and added this as a "fix".

What do you think? Happy to revert this if I misunderstood how it works.

Reply:
Thanks Adrien. Discussed offline, this is a very interesting use case so I added a unit test to cover it.

The existing implementation demonstrates a non-monotonic quantile function here, a big red flag. Reverting to the old version (matching 3.2) leads to inaccuracies when the digest contains only singleton centroids, but this is preferable and a simpler solution. The inaccuracies for the singleton-only case will be mitigated when we switch to a hybrid digest that will use a sorted array of values for small sample populations.

I applied your suggestion, resolving the thread. Please feel free to reopen or add further comments.